src/ralph/dns/views.py

```diff
@@ -22,37 +21,52 @@ def __init__(self, *args, **kwargs):
         self.dnsaas = DNSaaS()
         return super().__init__(*args, **kwargs)
 
-    def get_formset(self):
-        FormSet = formset_factory(  # noqa
-            DNSRecordForm, formset=BaseFormSet, extra=2,
-            can_delete=True
-        )
+    def get_forms(self):
+        forms = []
         initial = self.dnsaas.get_dns_records(
             self.object.ipaddress_set.all().values_list(
                 'address', flat=True
             )
         )
-        return FormSet(
-            data=self.request.POST or None,
-            initial=initial,
-        )
+        for item in initial:
+            forms.append(DNSRecordForm(item))
+
+        forms.append(DNSRecordForm())
+        return forms
 
     def get(self, request, *args, **kwargs):
-        kwargs['formset'] = self.get_formset()
+        kwargs['forms'] = self.get_forms()
         return super().get(request, *kwargs, **kwargs)
 
     def post(self, request, *args, **kwargs):
-        formset = self.get_formset()
-        if formset.is_valid():
-            for i, form in enumerate(formset.forms):
-                if form.cleaned_data.get('DELETE'):
-                    self.dnsaas.delete_dns_records(form.cleaned_data['pk'])
-                elif form.has_changed():
-                    if form.cleaned_data['pk']:
-                        self.dnsaas.update_dns_records(form.cleaned_data)
-                    else:
-                        self.dnsaas.create_dns_records(form.cleaned_data)
-            return HttpResponseRedirect('.')
-
-        kwargs['formset'] = formset
+        forms = self.get_forms()
+        current_form = DNSRecordForm(request.POST)
+        for i, form in enumerate(forms):
+            if (
+                str(form.data.get('pk', '')) ==
+                str(current_form.data.get('pk', ''))
+            ):
+                forms[i] = current_form
+                break
+
+        if current_form.is_valid():
+            if current_form.data.get('delete'):
+                errors = self.dnsaas.delete_dns_records(form.data['pk'])
+            elif current_form.cleaned_data.get('pk'):
+                errors = self.dnsaas.update_dns_records(
+                    current_form.cleaned_data
+                )
+            else:
+                errors = self.dnsaas.create_dns_records(
+                    current_form.cleaned_data
+                )
+
+            if errors:
+                for field_name, field_errors in errors.items():
```

Comment:
This assumes `field_errors` is `iterable`, does it work then with: https://github.com/allegro/ralph/pull/2311/files#diff-69d3de14f1ee658925a59b182b232129R139